docstring-formatter.lisp

```diff
@@ -0,0 +1,133 @@
+(in-package #:org.shirakumo.documentation-utils)
+
+#|
+Full protocol of docstring formatter.
+|#
+
+(defclass fundamental-docstring-formatter ()
+  ()
+  (:documentation "Fundamental, protocol class of docstring formatter."))
+
+
+(defclass fundamental-type-node ()
+  ()
+  (:documentation "Fundamental, protocol class of all type-nodes"))
+
+
+(defgeneric type-node-instance (generator type)
+  (:documentation "Return instance of type node corresponding to type."))
+
+
+(defgeneric visiting-order (generator node)
+  (:method ((generator fundamental-docstring-formatter)
+            (node fundamental-type-node))
+    (build-visiting-order generator node)))
+
+
+(defgeneric build-visiting-order (generator node)
+  (:method-combination list)
+  (:method list ((generator fundamental-docstring-formatter)
+                 (node fundamental-type-node))
+    '(:description :arguments :returns :examples
+      :side-effects :exceptional-situations :notes))
+  (:method :around ((generator fundamental-docstring-formatter)
+                    (node fundamental-type-node))
+    (apply #'append (reverse (call-next-method)))))
+
+
+(defgeneric visit-one (output data generator label node before)
+  (:method ((output stream)
+            (data (eql nil))
+            (generator fundamental-docstring-formatter)
+            (label symbol)
+            (node fundamental-type-node)
+            (before list))
+    nil))
+
+
+(defgeneric visit-all (output generator node form)
+  (:method ((output stream)
+            (generator fundamental-docstring-formatter)
+            (node fundamental-type-node)
+            (form string))
+    (print form output))
+  (:method ((output t)
+            (generator fundamental-docstring-formatter)
+            (node fundamental-type-node)
+            (form list))
+    (let ((order (visiting-order generator node))
+          (before nil))
+      (dolist (next order)
+        (let ((data (getf form next)))
+          (visit-one output
+                     data
+                     generator
+                     next
+                     node
+                     before)
+          (unless (null data)
+            (push next before)))))
+    output))
+
+(defclass basic-docstring-formatter (fundamental-docstring-formatter)
+  ())
+
+
+(let ((node (make-instance 'fundamental-type-node)))
+  (defmethod type-node-instance ((generator basic-docstring-formatter)
+                                 (type symbol))
+    node))
```

Comment:
This whole thing seems very suspect. Why is the type ignored? Why the lexical binding instead of just creating a new instance every call?